packages/react/src/ReactElement.js

```diff
@@ -164,6 +164,91 @@ const ReactElement = function(type, key, ref, self, source, owner, props) {
   return element;
 };
 
+/**
+ * https://github.com/reactjs/rfcs/pull/107
+ */
+export function jsx(type, config, maybeKey /*, source, self */) {
+  let propName;
+
+  // Reserved names are extracted
+  const props = {};
+
+  let key = null;
+  let ref = null;
+  let self = null;
+  let source = null;
+
+  // TODO: I don't allow null/undefined at dev time
+  // I think that decision makes sense because the goal is to pass props
+  // straight through. Is it React's responsibility to create an empty object
+  // if you pass null/undefined?
+  if (config != null) {
+    if (hasValidRef(config)) {
+      ref = config.ref;
+    }
+
+    if (hasValidKey(config)) {
+      key = '' + config.key;
+    }
+
+    // Remaining properties are added to a new props object
+    for (propName in config) {
+      if (
+        hasOwnProperty.call(config, propName) &&
+        !RESERVED_PROPS.hasOwnProperty(propName)
+      ) {
+        props[propName] = config[propName];
+      }
+    }
+  }
+
+  if (key === null && maybeKey !== undefined) {
+    key = maybeKey;
+  }
+
+  // Resolve default props
+  if (type && type.defaultProps) {
+    const defaultProps = type.defaultProps;
+    for (propName in defaultProps) {
+      if (props[propName] === undefined) {
+        props[propName] = defaultProps[propName];
+      }
+    }
+  }
+  if (__DEV__) {
+    if (key || ref) {
+      const displayName =
+        typeof type === 'function'
+          ? type.displayName || type.name || 'Unknown'
+          : type;
+      if (key) {
+        defineKeyPropWarningGetter(props, displayName);
+      }
+      if (ref) {
+        defineRefPropWarningGetter(props, displayName);
+      }
+    }
+
+    // these aren't explicitly defined because they are implementation details of dev time transform
+    // shouldn't really be exposed as part of the api. TODO: should they be?
+    if (arguments[3] !== undefined) {
```

Comment:
I think we probably should have a separate DEV only method for these because we want to ensure that it doesn't even work in prod so that nobody accidentally enables this in a prod environment. E.g. jsxDEV(...)

Additionally, we probably want to reserve the last arguments for the "flags" proposal.

Reply:
By flags proposal do you mean the static/not static flag? If we do that do the two functions (`jsx`/`jsxs`) still make sense?

`jsxDEV` is a bit tricky to write. But I'll put up changes to show how I'm imagining it.